heron/metricscachemgr/src/java/com/twitter/heron/metricscachemgr/metricscache/CacheCore.java

```diff
@@ -0,0 +1,543 @@
+                break;
+              case 1:
+                getMinuteMetrics(metricDatum.metricValue,
+                    request.startTime, request.endTime, bucketId, type);
+                break;
+              case 2:
+                getRawMetrics(metricDatum.metricValue,
+                    request.startTime, request.endTime, bucketId, type);
+                break;
+              default:
+                LOG.warning("unknown aggregationGranularity type "
+                    + request.aggregationGranularity);
+            }
+          } // end for: instance
+        } // end for: component
+      } // end for: metric
+      return response;
+    }
+  }
+
+  private void getRawMetrics(List<MetricsCacheQueryUtils.MetricTimeRangeValue> metricValue,
+                             long startTime, long endTime, long bucketId,
+                             MetricsFilter.MetricAggregationType type) {
+    LOG.info("getRawMetrics " + startTime + " " + endTime);
+    Long startKey = cacheMetric.floorKey(startTime);
+    for (Long key = startKey != null ? startKey : cacheMetric.firstKey();
+         key != null && key <= endTime;
+         key = cacheMetric.higherKey(key)) {
+      LinkedList<MetricDatapoint> bucket = cacheMetric.get(key).get(bucketId);
+
+      if (bucket != null) {
+        for (MetricDatapoint datapoint : bucket) {
+          if (startTime <= datapoint.timestamp && datapoint.timestamp <= endTime) {
+            // per data point
+            MetricsCacheQueryUtils.MetricTimeRangeValue rangeValue =
+                new MetricsCacheQueryUtils.MetricTimeRangeValue();
+            rangeValue.startTime = datapoint.timestamp;
+            rangeValue.endTime = datapoint.timestamp;
+            rangeValue.value = datapoint.value;
+            metricValue.add(rangeValue);
+          }
+        } // end bucket
+      }
+
+    } // end tree
+  }
+
+  // we assume the metric value is Double: compatible with tmaster
+  @SuppressWarnings("fallthrough")
+  private void getMinuteMetrics(List<MetricsCacheQueryUtils.MetricTimeRangeValue> metricValue,
+                                long startTime, long endTime, long bucketId,
+                                MetricsFilter.MetricAggregationType type) {
+    LOG.info("getMinuteMetrics " + startTime + " " + endTime);
+    Long startKey = cacheMetric.floorKey(startTime);
+    for (Long key = startKey != null ? startKey : cacheMetric.firstKey();
+         key != null && key <= endTime;
+         key = cacheMetric.higherKey(key)) {
+      LinkedList<MetricDatapoint> bucket = cacheMetric.get(key).get(bucketId);
+
+      if (bucket != null) {
+        // per bucket
+        long countAvg = 0;
+
+        MetricsCacheQueryUtils.MetricTimeRangeValue rangeValue =
+            new MetricsCacheQueryUtils.MetricTimeRangeValue();
+        rangeValue.startTime = Long.MAX_VALUE;
+        rangeValue.endTime = 0;
+
+        double result = 0;
+        for (MetricDatapoint datapoint : bucket) {
```

Comment:
A lot of this method looks the same as getAggregatedMetrics. Can you refactor for reuse?